### DISCUSSION TITLE
[FEATURE REQUEST] Junos Outbound SSH support

### DISCUSSION THREAD

**Discussion by natemellendorf** [saltstack/salt#67346] [Feature Requests]
**Is your feature request related to a problem? Please describe.**

Today, it appears that Salt supports the ability to establish an outbound SSH tunnel to a Junos device for remote management. (via Salt SSH)

This works well for devices that are publicly exposed or are privately available to the Salt master. However, devices behind a network address translation boundary (NAT) require exceptions to be made before a Salt master can communicate with them.

**Describe the solution you'd like**

A proposal, would be to configure an OpenSSH server on a Salt master. As Junos devices (SRX firewalls, EX / QFX switches) support outbound SSH, this would allow devices from behind a NAT boundary to phone home and connect to a centralized Salt master (or Syndic) for remote configuration. When devices connect, Salt could fire an event and run the device through applying states.

*Note: This is a fairly large, open ended feature request. One that I’m placing here to gauge desire, collaborate on, and see if it’s even possible / worth the investment.

**Describe alternatives you've considered**

Salt establishing an outbound SSH session to a device makes sense and it works. It’s familiar to many, and really isn’t something new. It also works across vendors, where this proposal would likely be Junos specific. 

The NAT issue proposed can be solved with firewall exceptions and port address translation (PAT) or static NAT entries to expose devices. (Granted, these doesn’t scale well and aren’t dynamic.)

**Additional context**

A good portion of this functionality is coded today and can be found here: https://github.com/jeremyschulman/junospyez-ossh-server

In summary (and from testing), a Junos device running current firmware can establish an outbound SSH session to this server. The server can then reach into the device and pull configuration data and make changes via NETCONF.

**SEP**

If this sounds like it needs a SEP, I would be more than happy to open one.

**Comment by Akm0d** [comment]
I wonder if this is something [heist](https://gitlab.com/saltstack/pop/heist) would solve well
